### PR TITLE
rrweb extension implementation

### PR DESCRIPTION
**Pull request by YunFeng0817**
### Features

- [x] Record web pages across tabs
- [x] Pause the recording and resume it at any time
- [x] Straightforward  action page

- Start recording button

image

   Stop / Pause recording buttons
    
image

- [x] A session list view to show all records

image

- [x] An embedded rrweb-player
image

- [x] Support [Manifest V3](https://developer.chrome.com/docs/extensions/mv3/intro/)
- [x] Support **Chromium** Based Browsers / **Firefox**
- [x] Support recording **cross-origin** iframes

### Storage

1. Recording status is stored in the Extension's local storage.
2. Recorded sessions are recorded in [indexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API).
3. Plan to put all settings in the Extension's sync storage.

### Next Version TODO

- [ ] Enable to config rrweb recording options in the extension's setting page
- [ ] Add an option to upload recorded sessions to remote storage

### CODE REVIEW

File: packages/web-extension/src/pages/Player.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { useRef, useEffect, useState } from 'react';
+import { useParams } from 'react-router-dom';
+import rrwebPlayer from 'rrweb-player';
+import Replayer from 'rrweb-player';
```
Comment (Juice10) [inline packages/web-extension/src/pages/Player.tsx:4]: Looks like your importing the same thing twice under a different name

File: packages/web-extension/src/pages/Player.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      .catch((err) => {
+        console.error(err);
+      });
+    getEvents(sessionId)
+      .then((events) => {
+        if (!playerElRef.current || !sessionId) return;
+
+        const linkEl = document.createElement('link');
+        linkEl.href =
+          'https://cdn.jsdelivr.net/npm/rrweb-player@1.0.0-alpha.3/dist/style.css';
```
Comment (Juice10) [inline packages/web-extension/src/pages/Player.tsx:35]:
```suggestion
          'https://cdn.jsdelivr.net/npm/rrweb-player@latest/dist/style.css';
```

File: packages/web-extension/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+export type RecordStoppedMessage = {
+  message: MessageName.RecordStopped;
+  events: eventWithTime[];
+  endTimestamp: number;
+  session?: Session;
+};
+
+export type HeartBreathMessage = {
```
Comment (Juice10) [inline packages/web-extension/src/types.ts:80]: Typo
```suggestion
export type HeartBeatMessage = {
```

File: packages/web-extension/src/content/index.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import type { eventWithTime } from '@rrweb/types';
+import {
+  LocalData,
+  LocalDataKey,
+  RecorderStatus,
+  ServiceName,
+  Session,
+  RecordStartedMessage,
+  RecordStoppedMessage,
+  HeartBreathMessage,
```
Comment (Juice10) [inline packages/web-extension/src/content/index.ts:12]:
```suggestion
  HeartBeatMessage,
```

File: packages/web-extension/src/content/index.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    });
+  });
+
+  window.addEventListener(
+    'message',
+    (event: {
+      data:
+        | RecordStartedMessage
+        | RecordStoppedMessage
+        | HeartBreathMessage
```
Comment (Juice10) [inline packages/web-extension/src/content/index.ts:83]:
```suggestion
        | HeartBeatMessage
```

File: packages/web-extension/src/content/index.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        clearRecorderCb?.();
+        clearRecorderCb = undefined;
+        stopResponseCb(newData);
+        void Browser.storage.local.set({
+          [LocalDataKey.bufferedEvents]: newData.events,
+        });
+      } else if (event.data.message === MessageName.HeartBeat)
+        void Browser.storage.local.set({
+          [LocalDataKey.bufferedEvents]: storedEvents.concat(
+            (event.data as HeartBreathMessage).events,
```
Comment (Juice10) [inline packages/web-extension/src/content/index.ts:114]:
```suggestion
            (event.data as HeartBeatMessage).events,
```

File: packages/web-extension/src/content/inject.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    }) || null;
+  window.postMessage({
+    message: MessageName.RecordStarted,
+    startTimestamp: Date.now(),
+  } as RecordStartedMessage);
+  setIntervalId = (setInterval(() => {
+    window.postMessage({
+      message: MessageName.HeartBeat,
+      events,
+    });
```
Comment (Juice10) [inline packages/web-extension/src/content/inject.ts:27]: It looks like this is going to constantly send the same events over and over again. I think we should probably clear the events array after we've sent the events

Comment (Mark-Fenng) [reply packages/web-extension/src/content/inject.ts:27]: Good point, I will improve this.

**Comment by Yuyz0112** [comment]
the indexeddb storage part looks very useful, hope we can move it into a separate re-usable package in this or future PRs.

File: packages/web-extension/package.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "web-extension",
```
Comment (Juice10) [inline packages/web-extension/package.json:2]: This is the name that will be used to publish the extension on NPM, we should probably change this to a namespaced name `@rrweb/web-extension`

Comment (Juice10) [inline packages/web-extension/package.json:2]:
```suggestion
  "name": "@rrweb/web-extension",
  "private": true,
```

File: .changeset/lovely-pears-cross.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+'web-extension': patch
```
Comment (Juice10) [inline .changeset/lovely-pears-cross.md:2]:
```suggestion
'@rrweb/web-extension': patch
```

**Review by Juice10 — APPROVED** [review]

**Pull request merged**